### DISCUSSION TITLE
On latency

### DISCUSSION THREAD

**Discussion by timlod** [mido/mido#485] [General]
Hi,
First off, thanks for building mido!
I'm building a looper which receives input via MIDI, and I've been using mido for that portion of it.
Next to just looping recordings, I also have functionality to trigger samples.

I was wondering what was the lowest-latency way of going from generating a MIDI signal to acting on it programmatically?
Currently I'm using RtMidi and a callback, and I think there is a stochastic delay somewhere on the order of <2ms (this is pure conjecture based on visually identifying the moment I pressed a key to generate a MIDI event on an audio waveform, and plotting when the program received the event).
I already `ports.set_sleep_time(0)`, which I believe in theory should allow MIDI polling as quickly as the system can interrupt (meaning switching to a PREEMPT_RT or similar kernel supposedly would reduce this latency).

I want to implement as-accurate-as-possible timing in my application, so any pointers to where I can read more about this would be welcome!

**Comment by rdoursenaud** [comment]
As far as I know, the timing information is provided by the MIDI interface driver from the hardware if available or using a system timer.
`RtMidi` has a property to retrieve this very accurate timestamp but we don't expose it yet.
I'm working on changes that should allow that but haven't finished yet.
See #241, #193, #370.
It might be what you're looking for if you need "recording" accuracy.

If you need to have the lowest latency possible for a "live" application, I believe that you're already using the best method: the callback.

> I already `ports.set_sleep_time(0)`, which I believe in theory should allow MIDI polling as quickly as the system can interrupt (meaning switching to a PREEMPT_RT or similar kernel supposedly would reduce this latency)

Since you're using a callback, you're not polling!
In this mode, `RtMidi` calls `mido` as soon as it receives a full (valid) message or chunk for long SysEx.
This seems superfluous to me.

**Reply by rdoursenaud** [reply]
I forgot to add that `mido`'s processing inherently adds a little bit of computational delay.
To achieve the lowest latency possible, you may want to use `python-rtmidi` directly if you don't require `mido` niceties.

**Reply by timlod** [reply]
Thanks a lot for the insight!
I misunderstood how the callback was implemented. I'll have a look at how things would change when using RtMidi directly and see how they compare!

**Reply by rdoursenaud** [reply]
You're welcome.
I'd love to read back from you and the results from your experiments.
It's an area of interest to me and could be beneficial information for the community.
Best of luck in your endeavors!

**Reply by timlod** [reply]
By checking the code I find that mido's overhead is mainly just `msg = Message.from_bytes(event[0])` (this is in _callback_wrapper in rtmidi.py).
I measured this by creating a `rtmidi` callback and measuring the time this takes manually.
On my machine, mido adds around 0.1ms of latency, on average.
